### PR TITLE
fix: fix eval bugs and behavior bugs

### PR DESCRIPTION
**Pull request by zhexuanxu**
### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/utils/metric_utils.py [hunk 1]
```diff
@@ -63,19 +63,24 @@ def compute_evaluate_metrics(eval_metrics_list):
...
         count = count_trajectories(eval_metrics)
         trajectory_counts.append(count)
 
     for env_info_key in env_info_keys:
-        all_eval_metrics[env_info_key] = [
-            eval_metrics[env_info_key] for eval_metrics in eval_metrics_list
+        metric = [
+            eval_metrics[env_info_key]
+            for eval_metrics in eval_metrics_list
+            if env_info_key in eval_metrics
```
Comment (qurakchin) [inline rlinf/utils/metric_utils.py:80]: In behavior, are there some keys may not in all env infos?

File: rlinf/workers/env/env_worker.py [hunk 2]
```diff
@@ -89,6 +89,8 @@ def __init__(self, cfg: DictConfig):
         )
         self.actor_split_num = self.get_actor_split_num()
 
+        self._prev_dones: list[torch.Tensor | None] = [None] * self.stage_num
```
Comment (qurakchin) [inline rlinf/workers/env/env_worker.py]: a [stage_num] tensor with initialized with zero is simpler.

Comment (qurakchin) [reply rlinf/workers/env/env_worker.py]: why using a dash prefix in this attr name?

File: rlinf/utils/metric_utils.py (hunk 1 above)
Comment (zhexuanxu) [reply rlinf/utils/metric_utils.py:80]: currently, the env info contains all keys

File: rlinf/workers/env/env_worker.py (hunk 2 above)
Comment (zhexuanxu) [reply rlinf/workers/env/env_worker.py]: done

**Review by qurakchin — APPROVED** [review]

File: rlinf/envs/behavior/behavior_env.py [hunk 3]
```diff
@@ -302,6 +305,13 @@ def chunk_step(self, chunk_actions):
             obs_list.append(self._wrap_obs(raw_obs_list[i]))
             infos_list.append(infos)
 
+        self._step_count += chunk_size
```
Comment (anHappyDog) [inline rlinf/envs/behavior/behavior_env.py]: The behavior already has the maximum number of steps set according to the configuration, so there is no need to add force_trunction here.

File: rlinf/envs/behavior/behavior_env.py [hunk 4]
```diff
@@ -361,7 +371,7 @@ def device(self):
 
     @property
     def elapsed_steps(self):
-        return torch.tensor(self.cfg.max_episode_steps)
+        return self._step_count
```
Comment (anHappyDog) [inline rlinf/envs/behavior/behavior_env.py]: Behavior will return `eposide_length` in info dict, maybe recording that is better.

File: rlinf/workers/env/env_worker.py [hunk 5]
```diff
@@ -89,6 +89,15 @@ def __init__(self, cfg: DictConfig):
         )
         self.actor_split_num = self.get_actor_split_num()
 
+        self.train_prev_done: list[torch.Tensor] = [
```
Comment (anHappyDog) [inline rlinf/workers/env/env_worker.py]: here `train_prev_done` is not used.

File: rlinf/envs/behavior/behavior_env.py [hunk 6]
```diff
@@ -426,6 +439,30 @@ def _record_metrics(self, rewards, infos):
         infos = {"episode": to_tensor(list_of_dict_to_dict_of_list(info_lists))}
         return infos
 
+    @staticmethod
+    def _extract_info_done(info: dict) -> bool:
```
Comment (anHappyDog) [inline rlinf/envs/behavior/behavior_env.py:424]: maybe we could simplify code here, `info` dict must have ["done"]["termination_conditions"], and every tc in info["done"]["termination_conditions"] must have key `done` and `success`, which are both bool variables.

File: rlinf/envs/behavior/behavior_env.py (hunk 3 above)
Comment (zhexuanxu) [reply rlinf/envs/behavior/behavior_env.py]: env can't see `max_episode_steps`, therefore I set a variable `_step_count`

File: rlinf/envs/behavior/behavior_env.py (hunk 4 above)
Comment (zhexuanxu) [reply rlinf/envs/behavior/behavior_env.py]: they are equivalent but I believe this way is more clear

File: rlinf/envs/behavior/behavior_env.py (hunk 6 above)
Comment (zhexuanxu) [reply rlinf/envs/behavior/behavior_env.py:424]: done

File: rlinf/workers/env/env_worker.py (hunk 5 above)
Comment (zhexuanxu) [reply rlinf/workers/env/env_worker.py]: it's will be used in main later, since there is a same issue

File: rlinf/envs/behavior/behavior_env.py (hunk 3 above)
Comment (anHappyDog) [reply rlinf/envs/behavior/behavior_env.py]: RLinf already set BEHAVIOR-1K's termination_config's sub attr `max_steps` to attr `max_episode_steps` in RLinf's config, so there is no need to record force termination. The added `_step_count` seems only used to help record force termination.  Also, in  `info` dict that every BEHAVIOR-1K step returns, it has a key `episode_length` that already record current length.

File: rlinf/envs/behavior/behavior_env.py [hunk 7]
```diff
@@ -405,13 +405,16 @@ def _record_metrics(self, rewards, infos):
...
-            if "fail" in info:
-                self.fail_once[env_idx] = self.fail_once[env_idx] | info["fail"]
+            done_dict = (
+                info.get("done", {}) if isinstance(info.get("done"), dict) else {}
+            )
+            step_success = done_dict.get("success", False) or info.get("success", False)
+            self.success_once[env_idx] = self.success_once[env_idx] | step_success
+            episode_info["success_once"] = self.success_once[env_idx].clone()
+
+            step_fail = done_dict.get("fail", False) or info.get("fail", False)
```
Comment (anHappyDog) [inline rlinf/envs/behavior/behavior_env.py]: the returned `info` dict does not contain key `fail`, and the `fail_once` metric is actually not used at all. Could you please remove fail_once and modify the `record_metric` function?

File: rlinf/envs/behavior/behavior_env.py [hunk 8]
```diff
@@ -405,13 +405,16 @@ def _record_metrics(self, rewards, infos):
...
             self.returns[env_idx] += reward
-            if "success" in info:
-                self.success_once[env_idx] = (
-                    self.success_once[env_idx] | info["success"]
-                )
-                episode_info["success_once"] = self.success_once[env_idx].clone()
-            if "fail" in info:
-                self.fail_once[env_idx] = self.fail_once[env_idx] | info["fail"]
+            done_dict = (
+                info.get("done", {}) if isinstance(info.get("done"), dict) else {}
```
Comment (anHappyDog) [inline rlinf/envs/behavior/behavior_env.py]: the returned `info` dict must have key `done`, and the corresponding value must be a dict. Also, it must contain key `success`, and it must be a bool variable. The code here could be simplified.

File: rlinf/utils/metric_utils.py [hunk 9]
```diff
@@ -63,19 +63,24 @@ def compute_evaluate_metrics(eval_metrics_list):
         dict: Aggregated metrics with mean values and trajectory count
     """
     all_eval_metrics = {}
-    env_info_keys = eval_metrics_list[0].keys()
+    env_info_keys: set[str] = set()
```
Comment (anHappyDog) [inline rlinf/utils/metric_utils.py:66]: Does the metrics list's element dict may have different keys？

Comment (zhexuanxu) [reply rlinf/utils/metric_utils.py:66]: the [0] may be none, while others have keys
it's checked with @guozhen1997 and @qurakchin

File: rlinf/envs/behavior/behavior_env.py (hunk 7 above)
Comment (zhexuanxu) [reply rlinf/envs/behavior/behavior_env.py]: ok

File: rlinf/envs/behavior/behavior_env.py (hunk 8 above)
Comment (zhexuanxu) [reply rlinf/envs/behavior/behavior_env.py]: ok

File: rlinf/envs/behavior/behavior_env.py (hunk 3 above)
Comment (zhexuanxu) [reply rlinf/envs/behavior/behavior_env.py]: fixed

**Review by anHappyDog — APPROVED** [review]

**Review by Lin-xs — APPROVED** [review]

**Review by guozhen1997 — APPROVED** [review]
